### PR TITLE
Clip Focus Adorners

### LINKED ISSUE #337: Focus adorner for ListBox item is rendered over other controls

### PR DESCRIPTION
**Pull request by grokys**
Focus adorners were not being clipped to the adorned controls clip bounds. This fixes that by adding a `Clip` to the adorner.

Fixes #337
Depends on #1349

### CODE REVIEW

File: src/Avalonia.Themes.Default/FocusAdorner.xaml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <Style xmlns="https://github.com/avaloniaui" Selector=":is(Control)">
   <Setter Property="FocusAdorner">
     <FocusAdornerTemplate>
-      <Rectangle Stroke="Black"
```
Comment (jkoritzinsky) [inline src/Avalonia.Themes.Default/FocusAdorner.xaml]: Can we pick a more descriptive name than "abcd"?

File: src/Avalonia.Visuals/Rendering/DeferredRenderer.cs [hunk 2]
```diff
@@ -243,6 +243,15 @@ private void Render(IDrawingContextImpl context, VisualNode node, IVisual layer,
             {
                 clipBounds = node.ClipBounds.Intersect(clipBounds);
 
+                var np = node.Visual.GetType().GetProperty("Name");
+                var name = (string)np.GetValue(node.Visual);
+
+                if (name == "abcd")
```
Comment (jkoritzinsky) [inline src/Avalonia.Visuals/Rendering/DeferredRenderer.cs]: Can we pick a more descriptive name (and maybe save it in a constant so it isn't a magic string)?
Also, is this used for anything other than debugging?

**Review by jkoritzinsky — CHANGES_REQUESTED** [review]
Most of this looks good, except for what looks like some debugging code from development that should be either updated with useful naming or removed.

Although, I'm having trouble discerning why a lot of the refactoring of the `Geometry` classes are in this PR. Are there additional refactorings here in comparison to #1349, or a subset or something else?

File: src/Avalonia.Visuals/Rendering/DeferredRenderer.cs (hunk 2 above)
Comment (grokys) [reply src/Avalonia.Visuals/Rendering/DeferredRenderer.cs]: Oh good catch, this is debug code.

**Comment by grokys** [comment]
Ah yeah, this should have WIP in the title - I was working on this and decided that I needed to do the Geometry work first. I need to update this PR with the changes from that PR and remove the debug code.

**Comment by grokys** [comment]
Ok, removed the debug code and updated from #1349 - once that is merged you should see the changes from just this PR. No longer WIP.

**Review by jkoritzinsky — APPROVED** [review]
LGTM

**Pull request merged**